[WIP] Consolidated 4.10 RAN docs (#43890)
openshift/openshift-docs: modules/cnf-topology-aware-lifecycle-operator-platform-update.adoc

  sourceFiles: 
    - fileName: ImageSignature.yaml <1>
      policyName: "platform-upgrade-prep"
      binaryData:
        ${DIGEST_ALGO}-${DIGEST_ENCODED}: ${SIGNATRUE_BASE64} <2>
    - fileName: DisconnectedICSP.yaml
      policyName: "platform-upgrade-prep"
      metadata:
        name: disconnected-internal-icsp-for-ocp
      spec:
        repositoryDigestMirrors:
          - mirrors:
              - registry.example.com:5000/openshift4 <3>
            source: quay.io/openshift-release-dev/ocp-release <4>
    - fileName: ClusterVersion.yaml <5>
      policyName: "platform-upgrade-prep"
      metadata:
        name: version
      spec:
        upstream: http://upgrade.example.com/images/upgrade-graph_stable-4.10
    - fileName: ClusterVersion.yaml <6>
      policyName: "platform-upgrade"
      metadata:
        name: version
      spec:
        channel: "stable-4.10"
        upstream: http://upgrade.example.com/images/upgrade-graph_stable-4.10
        desiredUpdate:
          version: 4.10.4
      status:
        history:
          - version: 4.10.4
            state: "Completed"
----
<1> The `ConfigMap` CR contains the signature of the desired release image to update to.

> <2> Shows the image signature of the desired {product-title} release. To get the signature and replace the variables with the values, see step 2-7 of xref:../updating/updating-restricted-network-cluster.html#update-configuring-image-signature[Creating an image signature config map manually].

Reviewer: This link doesn't have the steps of getting the image signature. Actually, the steps are only available in the OCP4.7 doc.
Instead of referencing to that link, in the **Preparing for the updates**, we should mention where/how to get the image signature and that is for later use in PGT for the platform upgrade. 
If the mirroring steps for **the mirror host does not have internet access** have been followed in **Mirroring the OpenShift Container Platform image repository** from https://deploy-preview-43890--osdocs.netlify.app/openshift-enterprise/latest/scalability_and_performance/ztp-deploying-disconnected.html#installation-mirror-repository_ztp-deploying-disconnected, the image signature can be found from ${REMOVABLE_MEDIA_PATH}/mirror/config/<image_signature_file>
Also document the manual steps as a second option.

Author: @Missxiaoguo I updated this part, please check if that's what you meant.

      policyName: "platform-upgrade-prep"
      binaryData:
        ${DIGEST_ALGO}-${DIGEST_ENCODED}: ${SIGNATRUE_BASE64} <2>
    - fileName: DisconnectedICSP.yaml
      policyName: "platform-upgrade-prep"
      metadata:
        name: disconnected-internal-icsp-for-ocp
      spec:
        repositoryDigestMirrors: <3>
          - mirrors:
            - quay-intern.example.com/ocp4/openshift-release-dev
            source: quay.io/openshift-release-dev/ocp-release
          - mirrors:
            - quay-intern.example.com/ocp4/openshift-release-dev
            source: quay.io/openshift-release-dev/ocp-v4.0-art-dev <4>
    - fileName: ClusterVersion.yaml <5>
      policyName: "platform-upgrade-prep"
      metadata:
        name: version
      spec:
        upstream: http://upgrade.example.com/images/upgrade-graph_stable-4.10
    - fileName: ClusterVersion.yaml <6>
      policyName: "platform-upgrade"
      metadata:
        name: version
      spec:
        channel: "stable-4.10"
        upstream: http://upgrade.example.com/images/upgrade-graph_stable-4.10
        desiredUpdate:
          version: 4.10.4
      status:
        history:
          - version: 4.10.4

>             state: "Completed"

Reviewer 2: Should be removed?

>       status:
>         history:
>           - version: 4.10.4
>             state: "Completed"

Reviewer 2: ```suggestion
```

----
<1> The `ConfigMap` CR contains the signature of the desired release image to update to.
<2> Shows the image signature of the desired {product-title} release. If you followed the steps in "Mirroring the OpenShift Container Platform image repository" section, you can get the image signature at `${REMOVABLE_MEDIA_PATH}/mirror/config/<image_signature_file>`.
<3> Shows the mirror repository that contains the desired {product-title} image.

> <4> Shows the source repository that the desired {product-title} image is pulled from. You can list the required mirror registries by running the `oc adm release mirror` command.

Reviewer: Please check out this part I updated in the Google doc.

    maxConcurrency: 1
  enable: true
----

.. Apply the policy to the Hub cluster by running the following command:
+
[source,terminal]
----
$ oc apply -f cgu-platform-upgrade-prep.yml
----

.. Monitor the update process and wait for completion. Then ensure that the policy becomes compliant by running the following command:
+
[source,terminal]
----
$ oc get policies --all-namespaces
----

. Pre-cache the images for the platform update.
.. Save the content of the platform update `ClusterGroupUpgrade` CR with the `du-upgrade-platform-upgrade` policy and the target clusters to the `cgu-platform-upgrade.yml` file. This `ClusterGroupUpgrade` CR is configured to only perform the pre-caching job without the actual cluster update.
+
[source,yaml]
----
apiVersion: ran.openshift.io/v1alpha1
kind: ClusterGroupUpgrade
metadata:
  name: cgu-platform-upgrade
  namespace: default
spec:
  managedPolicies:
  - du-upgrade-platform-upgrade
  preCaching: true
  clusters:
  - spoke1
  remediationStrategy:
    maxConcurrency: 1
  enable: false
----

.. Apply the CR to the Hub cluster by running the following command:
+
[source,terminal]
----
$ oc apply -f cgu-platform-upgrade-prep.yml
----

> .. Monitor the date process and wait for the pre-caching to complete. Check the status of pre-caching by running the following command on the spoke cluster:

Reviewer 2: ```suggestion
.. Monitor the update process and wait for the pre-caching to complete. Check the status of pre-caching by running the following command on the hub cluster:
```

+
[source,terminal]

> ----
> $ oc get jobs,pods -n openshift-talo-pre-cache
> ----

Reviewer 2: ```suggestion
----
$ oc get cgu cgu-platform-upgrade -o jsonpath='{.status.precaching.status}'
----
```
This will list the pre-caching status per cluster.

.Prerequisites

* Install the Topology Aware Lifecycle Operator.
* Update ZTP to the latest version. 
* Provision one or more managed clusters with ZTP.
* Mirror the desired image repository.
* Log in as a user with `cluster-admin` privileges.
* Create {rh-rhacm} policies in the Hub cluster.

.Procedure

. Create a `PolicyGenTemplate` CR for the platform update:
.. Save the following contents of the `PolicyGenTemplate` CR in the `du-upgrade.yaml` file.
+
.Example of `PolicyGenTemplate` for platform update 
+
[source,yaml]
----
apiVersion: ran.openshift.io/v1
kind: PolicyGenTemplate
metadata:
  name: "du-upgrade"
  namespace: "ztp-group-du-sno"
spec:
  bindingRules:
    group-du-sno: ""
  mcp: "master"
  remediationAction: inform
  sourceFiles: <1>
    - fileName: ImageSignature.yaml
      policyName: "platform-upgrade-prep"
      binaryData:
        ${DIGEST_ALGO}-${DIGEST_ENCODED}: ${SIGNATRUE_BASE64} <2>
    - fileName: DisconnectedICSP.yaml
      policyName: "platform-upgrade-prep"
      metadata:
        name: disconnected-internal-icsp-for-ocp
      spec:
        repositoryDigestMirrors:
          - mirrors:
              - registry.example.com:5000/openshift4 <3>

>             source: quay.io/openshift-release-dev/ocp-release <4>

Author: @yliu127 Updated with the help of Angie. Please check if correct

>             source: quay.io/openshift-release-dev/ocp-release <4>

Reviewer 3: This is still missing another item. Following two sources should be added. (please change the mirror)
 
```
- mirrors:
  - quay-intern.example.com/ocp4/openshift-release-dev
  source: quay.io/openshift-release-dev/ocp-release
- mirrors:
  - quay-intern.example.com/ocp4/openshift-release-dev
  source: quay.io/openshift-release-dev/ocp-v4.0-art-dev
```

// Module included in the following assemblies:
// Epic CNF-2600 (CNF-2133) (4.10), Story TELCODOCS-285
// * scalability_and_performance/ztp-deploying-disconnected.adoc

:_content-type: PROCEDURE
[id="talo-platform-update_{context}"]
= Performing a platform update

You can perform a platform update with the TALO.

.Prerequisites

* Install the Topology Aware Lifecycle Operator.
* Update ZTP to the latest version. 
* Provision one or more managed clusters with ZTP.
* Mirror the desired image repository.
* Log in as a user with `cluster-admin` privileges.
* Create {rh-rhacm} policies in the hub cluster.

.Procedure

. Create a `PolicyGenTemplate` CR for the platform update:
.. Save the following contents of the `PolicyGenTemplate` CR in the `du-upgrade.yaml` file.
+
.Example of `PolicyGenTemplate` for platform update 
+
[source,yaml]
----
apiVersion: ran.openshift.io/v1
kind: PolicyGenTemplate
metadata:
  name: "du-upgrade"
  namespace: "ztp-group-du-sno"
spec:
  bindingRules:
    group-du-sno: ""
  mcp: "master"
  remediationAction: inform
  sourceFiles: 
    - fileName: ImageSignature.yaml <1>
      policyName: "platform-upgrade-prep"
      binaryData:

>         ${DIGEST_ALGO}-${DIGEST_ENCODED}: ${SIGNATRUE_BASE64} <2>

Reviewer 4: Typo...
```suggestion
        ${DIGEST_ALGO}-${DIGEST_ENCODED}: ${SIGNATURE_BASE64} <2>
```

    - fileName: DisconnectedICSP.yaml
      policyName: "platform-upgrade-prep"
      metadata:
        name: disconnected-internal-icsp-for-ocp
      spec:
        repositoryDigestMirrors: <3>
          - mirrors:
            - quay-intern.example.com/ocp4/openshift-release-dev
            source: quay.io/openshift-release-dev/ocp-release
          - mirrors:
            - quay-intern.example.com/ocp4/openshift-release-dev
            source: quay.io/openshift-release-dev/ocp-v4.0-art-dev
    - fileName: ClusterVersion.yaml <4>
      policyName: "platform-upgrade-prep"

>       metadata:
>         name: version

Reviewer 4: Missing required annotation:
```suggestion
      metadata:
        name: version
        annotations:
          ran.openshift.io/ztp-deploy-wave: "1"
```

            source: quay.io/openshift-release-dev/ocp-v4.0-art-dev <4>
    - fileName: ClusterVersion.yaml <5>
      policyName: "platform-upgrade-prep"
      metadata:
        name: version
      spec:
        upstream: http://upgrade.example.com/images/upgrade-graph_stable-4.10
    - fileName: ClusterVersion.yaml <6>
      policyName: "platform-upgrade"
      metadata:
        name: version
      spec:
        channel: "stable-4.10"
        upstream: http://upgrade.example.com/images/upgrade-graph_stable-4.10
        desiredUpdate:
          version: 4.10.4
      status:
        history:
          - version: 4.10.4
            state: "Completed"
----
<1> The `ConfigMap` CR contains the signature of the desired release image to update to.
<2> Shows the image signature of the desired {product-title} release. Get the signature from the `checksum-${OCP_RELASE_NUMBER}.yaml` file you saved when following the procedures in the "Setting up the environment" section.
<3> Shows the mirror repository that contains the desired {product-title} image. Get the mirrors from the `imageContentSources.yaml` you saved when following the procedures in the "Setting up the environment" section.

> <4> Shows the source repository that the desired {product-title} image is pulled from. You should use the `imageContentSources` listed in the `oc adm release mirror` command output here.

Reviewer: <4> should be removed as it's already explained in <3>

+
.Example of `PolicyGenTemplate` for platform update 
+
[source,yaml]
----
apiVersion: ran.openshift.io/v1
kind: PolicyGenTemplate
metadata:
  name: "du-upgrade"
  namespace: "ztp-group-du-sno"
spec:
  bindingRules:
    group-du-sno: ""
  mcp: "master"
  remediationAction: inform
  sourceFiles: 
    - fileName: ImageSignature.yaml <1>
      policyName: "platform-upgrade-prep"
      binaryData:
        ${DIGEST_ALGO}-${DIGEST_ENCODED}: ${SIGNATURE_BASE64} <2>
    - fileName: DisconnectedICSP.yaml
      policyName: "platform-upgrade-prep"
      metadata:
        name: disconnected-internal-icsp-for-ocp
      spec:
        repositoryDigestMirrors: <3>
          - mirrors:
            - quay-intern.example.com/ocp4/openshift-release-dev
            source: quay.io/openshift-release-dev/ocp-release
          - mirrors:
            - quay-intern.example.com/ocp4/openshift-release-dev
            source: quay.io/openshift-release-dev/ocp-v4.0-art-dev
    - fileName: ClusterVersion.yaml <4>
      policyName: "platform-upgrade-prep"
      metadata:
        name: version
        annotations:
          ran.openshift.io/ztp-deploy-wave: "1"

>       spec:
>         upstream: http://upgrade.example.com/images/upgrade-graph_stable-4.10

Reviewer: ```suggestion
      spec:
        channel: "stable-4.10"
        upstream: http://upgrade.example.com/images/upgrade-graph_stable-4.10
```
